operate/webapp/src/main/java/io/camunda/operate/webapp/zeebe/operation/adapter/ServicesBasedAdapter.java

```diff
@@ -0,0 +1,380 @@
+                operationId));
+  }
+
+  @Override
+  public void cancelProcessInstance(final long processInstanceKey, final String operationId) {
+    executeCamundaServiceAnonymously(
+        (authentication) ->
+            withOperationReference(
+                operationReference ->
+                    processInstanceServices.cancelProcessInstance(
+                        new ProcessInstanceCancelRequest(processInstanceKey, operationReference)),
+                operationId));
+  }
+
+  @Override
+  public void updateJobRetries(final long jobKey, final int retries, final String operationId) {
+    // operationId is not used in the updateJob service method
+    executeCamundaServiceAnonymously(
+        (authentication) -> jobServices.updateJob(jobKey, new UpdateJobChangeset(retries, null)));
+  }
+
+  @Override
+  public void resolveIncident(final long incidentKey, final String operationId) {
+    // operationId is not used in the updateJob service method
+    executeCamundaServiceAnonymously(
+        (authentication) -> incidentServices.resolveIncident(incidentKey));
+  }
+
+  @Override
+  public long setVariables(
+      final long scopeKey,
+      final Map<String, Object> variables,
+      final boolean local,
+      final String operationId) {
+    final var variableDocumentRecord =
+        // operationId is not used in the updateJob service method
+        executeCamundaServiceAnonymously(
+            (authentication) ->
+                withOperationReference(
+                    operationReference ->
+                        elementInstanceServices.setVariables(
+                            new SetVariablesRequest(
+                                scopeKey, variables, local, operationReference)),
+                    operationId));
+    return variableDocumentRecord.getScopeKey();
+  }
+
+  protected static <T> CompletableFuture<T> withOperationReference(
+      final Function<Long, CompletableFuture<T>> command, final String id) {
+    try {
+      final long operationReference = Long.parseLong(id);
+      return command.apply(operationReference);
+    } catch (final NumberFormatException e) {
+      LOGGER.debug(
+          "The operation reference provided is not a number: {}. Ignoring propagating it to zeebe commands.",
+          id);
+    }
+    return null;
+  }
+
+  private <T> T executeCamundaServiceAnonymously(
+      final Function<Authentication, CompletableFuture<T>> method) {
+    return executeCamundaService(method, RequestMapper.getAnonymousAuthentication());
+  }
+
+  private <T> T executeCamundaService(
+      final Function<Authentication, CompletableFuture<T>> method,
+      final Authentication authentication) {
+    try {
+      return method.apply(authentication).join();
+    } catch (final Exception e) {
+      throw handleException(e);
+    }
+  }
+
+  private RuntimeException handleException(final Throwable error) {
+    return switch (error) {
```

Comment:
🔧 It needs to handle some "retriable" exception explicitly:

```
      Arrays.asList(
          Status.UNAVAILABLE.getCode(),
          Status.RESOURCE_EXHAUSTED.getCode(),
          Status.DEADLINE_EXCEEDED.getCode());
```
(cf. [link](https://github.com/camunda/camunda/blob/main/operate/webapp/src/main/java/io/camunda/operate/webapp/zeebe/operation/AbstractOperationHandler.java#L31-L35))

Reply:
Both adapters now have a `isExceptionRetriable` method which is used in the `AbstactOperationHandler`. So the retry logic remains the same, but the conditions change depending on which adapter is used.

Reply:
👍